sound/soc/sof/intel/hda-stream.c

```diff
@@ -193,23 +202,40 @@ int hda_dsp_stream_put(struct snd_sof_dev *sdev, int direction, int stream_tag)
 {
 	struct hdac_bus *bus = sof_to_bus(sdev);
 	struct hdac_stream *s;
+	bool active_capture_stream = false;
+	bool found = false;
 
 	spin_lock_irq(&bus->reg_lock);
 
-	/* find used stream */
+	/*
+	 * close stream matching the stream tag
+	 * and check if there are any open capture streams.
+	 */
 	list_for_each_entry(s, &bus->stream_list, list) {
-		if (s->direction == direction &&
-		    s->opened && s->stream_tag == stream_tag) {
+		if (!s->opened)
+			continue;
+
+		if (s->direction == direction && s->stream_tag == stream_tag) {
 			s->opened = false;
-			spin_unlock_irq(&bus->reg_lock);
-			return 0;
+			found = true;
+		} else if (s->direction == SNDRV_PCM_STREAM_CAPTURE) {
+			active_capture_stream = true;
 		}
 	}
 
 	spin_unlock_irq(&bus->reg_lock);
 
-	dev_dbg(sdev->dev, "stream_tag %d not opened!\n", stream_tag);
-	return -ENODEV;
+	/* Enable L1 support if there are no active capture streams */
+	if (!active_capture_stream)
+		snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, HDA_VS_EM2,
+					HDA_VS_EM2_L1SEN, HDA_VS_EM2_L1SEN);
```

Comment:
Similar as above, we might check and set this multiple times, even when we are putting playback stream, it is not efficiency to me.

Reply:
@keyonjie no thats not correct. update_bits will only write if the status changes. If the L1 is already disabled or enabled, it wont do anything. I dont think it is efficient to keep count of open streams at all. Just making sure there are no open streams here is quite efficient IMO

Reply:
Besides the thing to keep in mind is that this is just a workaround. I don't think there is any use to keeping a count of open streams once the workaround is removed

Reply:
Besides the thing to keep in mind is that this is just a workaround. I don't think there is any use to keeping a count of open streams once the workaround is removed

Reply:
Besides the thing to keep in mind is that this is just a workaround. I don't think there is any use to keeping a count of open streams once the workaround is removed

Reply:
> @keyonjie no thats not correct. update_bits will only write if the status changes. If the L1 is already disabled or 

OK, so the snd_sof_dsp_update_bits() has logic inside to prevent setting the same value multiple times.

> enabled, it wont do anything. I dont think it is efficient to keep count of open streams at all. Just making sure there are no open streams here is quite efficient IMO

The low efficient part in the PR you provided here to me is that we need to traverse the whole list to mark the active_capture_stream flag every time(previously we will return once the stream is found), letting alone that the status of capture streams actually won't change when we are *putting* playback streams.

Reply:
> Besides the thing to keep in mind is that this is just a workaround. I don't think there is any use to keeping a count of open streams once the workaround is removed

Agree about that it is just a workaround, and yes that's why I am proposing usage of count, removing a count and it's usage is more easier than pieces of code like it here IMO, removing the former one is error-prone to me.

Reply:
@keyonjie Hmm, I'd actually prefer not to add a refcount field. The existing code already has to loop over streams, so the extra cost of counting capture streams is minimal. And now this workaround is isolated to hda-stream.c. If we add refcounting, then we have a field in sof-priv.h that  requires explaining&understanding. I'd prefer to keep this as localized as possible.

Reply:
@keyonjie I have a similar opinion to @ranj063 and @kv2019i - reading the register, finding out, that the correct value is already set and then avoiding the write seems  a good enough approach here to me too.

Reply:
> @keyonjie Hmm, I'd actually prefer not to add a refcount field. The existing code already has to loop over streams, so the extra cost of counting capture streams is minimal. And now this workaround is isolated to hda-stream.c. If we add refcounting, then we have a field in sof-priv.h that requires explaining&understanding. I'd prefer to keep this as localized as possible.

@kv2019i what I proposed is actually localized, it is in struct sof_intel_hda_dev in hda.h, not sof-priv.h.

```
diff --git a/sound/soc/sof/intel/hda.h b/sound/soc/sof/intel/hda.h
index 73d7cc08afc2..f2dd68000575 100644
--- a/sound/soc/sof/intel/hda.h
+++ b/sound/soc/sof/intel/hda.h
@@ -394,6 +394,9 @@ struct sof_intel_hda_dev {
        /* the maximum number of streams (playback + capture) supported */
        u32 stream_max;
 
+       /* to store the count of opened capture stream, used for L1 support */
+       refcount_t capture_cnt;
+
        int irq;
 
        /* DMIC device */
```

Reply:
@keyonjie wrote
> @kv2019i what I proposed is actually localized, it is in struct sof_intel_hda_dev in hda.h, 
> not sof-priv.h.

... especially with the inline comment you put, this is better, agreed, but even hda.h exposes 'capture_cnt' to much more code. Someone learning the SOF HDA codebase, will be looking at hda.h structs and fields like 'capture_cnt' (that are only used for a very specific purpose and only by a few functions of the HDA code) add to the complexity of the interface. I agree if the computation cost is high enough, then having redundant state cached in structs is ok of course, but I don't think that's the case here.